accepted/00000-replace-salt-toaster.md

```diff
@@ -0,0 +1,192 @@
+testing Salt. Additionally, we have knowledge about running a test suite within
+our team that we want to take advantage of.
+
+# Detailed design
+[design]: #detailed-design
+
+<!-- This is the bulk of the RFC. Explain the design in enough detail for somebody familiar with the product to understand, and for somebody familiar with the internals to implement. -->
+
+<!-- This section should cover architecture aspects and the rationale behind disruptive technical decisions (when applicable), as well as corner-cases and warnings. Whenever the new feature creates new user interactions, this section should include examples of how the feature will be used. -->
+
+## Overview
+
+The Salt upstream test suite is executed in virtual machines (VMs) that run the
+different operating systems we support. These VMs are set up with sumaform and
+salt-jenkins. Terracumber triggers terraform to create the VMs from within a
+Jenkins job and launches the test suite, connecting to each VM over SSH.
+
+## Provisioning Test VMs with Sumaform + Salt-Jenkins
+
+A new module "salt\_testenv" is added to sumaform. This module defines a new
+role, "salt\_testenv", and the default values needed to run all test suites, e.g.
+4GiB of RAM. Sumaform uses this role to apply the `golden-images-provision.sls`
+state file from the Salt-Jenkins repository.
+
+Up to now, all state files used by sumaform are maintained in the sumaform Git
+repository. For provisioning the salt\_testenv VMs, we don't want to maintain
+our own states that define which dependencies need to be installed on what
+operating system. Salt upstream already maintains such states for their own test
+infrastructure. We re-use their states by adding `salt-jenkins` as a Git
+submodule to sumaform and merge the salt state trees.
+
+In order to avoid future state id clashes, all existing state ids will get a
+"sumaform-" prefix.
+
+### Merging Sumaform's and Salt-Jenkins' Salt Trees
+
+The state trees are merged by creating symbolic links inside sumaform's `salt`
+directory. These symbolic links point to each file in `salt-jenkins/state-tree`
+directory. For directories that exist in both tree, the files inside the
+respective directory in `salt-jenkin/state-tree` are linked instead.
+
+This symbolic link approach has the advantage that it clearly shows where each
+state file is maintained and can be cleaned up easily. Terraform's `file`
+provider, that sumaform uses to upload the `salt/` directory to each VM for
+provisioning can't handle symbolic links well. A
+[workaround](https://github.com/hashicorp/terraform/issues/13202#issuecomment-321578202)
+is possible, but it will touch a core sumaform mechanism.
+
+#### Alternatives
+
+Alternatively, we can copy the files instead. This loses the information where a
+state file is maintained and requires manually updating the copies whenever the
+used submodule revision is updated. Hard links would solve the last issue, I
```

Comment:
How about having Jenkins copy trees over to the sumaform directory, and then allowing sumaform to copy them over to the VM in turn? Is that what you are proposing here?

Reply:
Both the symbolic link approach and the copy approach to the state tree merger must be executed before running `terraform apply`. I haven't thought too hard about where to define the merging. I wouldn't hard-code it into a Jenkins job, at the very least it should be in a shell script to maintain the code properly.
 My preferred way, that should work based on my (limited) experience with it, is to define the merging in terracumber (symbolic links, hard links or copying). This area needs a bit more research.